src/AI/WallyAI.java

```diff
@@ -579,314 +1066,737 @@ public GameAction getNextAction(PriorityQueue<Unit> unitQueue, GameMap gameMap)
+        goals.add(coord);
       }
     }
 
-    if( goals.isEmpty() ) // Send 'em at production facilities if they haven't got anything better to do
+    if( goals.isEmpty() ) // Send 'em at production facilities if there's nothing to shoot
     {
       for( XYCoord coord : futureCapTargets )
       {
         MapLocation loc = gameMap.getLocation(coord);
-        if( unit.CO.unitProductionByTerrain.containsKey(loc.getEnvironment().terrainType)
-            && myArmy.isEnemy(loc.getOwner()) )
+        PathCalcParams pcp = new PathCalcParams(uc, gameMap).setTheoretical();
+        if( um.co.unitProductionByTerrain.containsKey(loc.getEnvironment().terrainType)
+            && myArmy.isEnemy(loc.getOwner())
+            && null != pcp.findShortestPath(coord) )
         {
           goals.add(coord);
         }
       }
     }
 
-    if( goals.isEmpty() ) // If there's really nothing to do, go to MY HQ
-      goals.addAll(myArmy.HQLocations);
-
-    Utils.sortLocationsByDistance(new XYCoord(unit.x, unit.y), goals);
     return goals;
   }
 
-  /** Functions as working memory to prevent eviction cycles */
-  private transient Set<Unit> evictionStack;
-  private static final int    EVICTION_STACK_MAX_DEPTH = 7;
+  private static class EvictionContext
+  {
+    AIModule whodunit;
+    GameMap map;
+    @SuppressWarnings("unused")
+    UnitPrediction[][] mapPlan;
+    int maxEvictionDepth;
+    private Stack<Unit> evictionStack;
+    private Stack<ActionPlan> postrequisites;
+    public EvictionContext(AIModule whodunit, GameMap gameMap, UnitPrediction[][] mapPlan, int evictionDepth, Unit toMove)
+    {
+      this.whodunit = whodunit;
+      this.map      = gameMap;
+      this.mapPlan  = mapPlan;
+      maxEvictionDepth = evictionDepth;
+      this.evictionStack = new Stack<>();
+      postrequisites = new Stack<>();
+      ActionPlan evictingPlan = mapPlan[toMove.x][toMove.y].toAchieve;
+      // If the action is my own action, don't ban me from considering it again
+      if( null != evictingPlan && toMove != evictingPlan.actor.unit )
+        enqueuePostReqPlans(mapPlan, evictingPlan, postrequisites);
+    }
+    public void push(Unit unit, ActionPlan plan)
+    {
+      evictionStack.push(unit);
+      postrequisites.push(plan);
+    }
+    public void pop()
+    {
+      evictionStack.pop();
+      postrequisites.pop();
+    }
+    public int getEvictionValue()
+    {
+      int value = 0;
+      for( ActionPlan plan : postrequisites )
+      {
+        value += plan.score;
+      }
+      return value;
+    }
+    public int remainingDepth()
+    {
+      return maxEvictionDepth - evictionStack.size();
+    }
+    /** Find the tiles a unit on startTile shouldn't go, based on any planned actions that need that tile to be open. */
+    public ArrayList<XYCoord> calcPostReqBans()
+    {
+      var bannedTiles = new ArrayList<XYCoord>();
+
+      for( ActionPlan plan : postrequisites )
+      {
+        var tile = new XYCoord(plan.actor.unit);
```

Comment:
This used to be `var tile = plan.actor.coord`. How is this different? Is the unit at a different coordinate than itself?

Reply:
There are (at least) 3 places the "unit's coordinate" is stored.
- The unit's actual coordinate. Wally cares about this some of the time, but not all of the time.
- The coordinate in `mapPlan` that the `ActionPlan` is stored (i.e. where the unit will move to complete its planned action this turn). Wally almost always cares about this one.
- The coordinate in the `ActionPlan`'s actor `UnitContext`. This should match the above.

It's pretty convoluted, but I don't think there's a cleaner way to enable pre-planning like this - It's kind of baked into the notion of considering "where I am" and "where I want to be" for your entire army at once.

Reply:
> The coordinate in the ActionPlan's actor UnitContext. This should match the above.

Match _which_ of the above? The first two don't sound like they should match.

> considering "where I am" and "where I want to be" for your entire army at once

This problem statement would seem to imply you need two stored values, not three.

Reply:
The coordinate where the `ActionPlan` is stored.
The `UnitContext` gets its coordinate set (to the end of the path) when I feed it the path it will use for its action.
And yes, the first two usually don't match (but units can act in place, so they *can* match).
> This problem statement would seem to imply you need two stored values, not three.

You're quite correct, but I also want to be able to get at the coordinate without having to iterate through the entire `mapPlan` looking for an identity match.